ALSA/ASoC: Change location where HDAudio stream pointer is stored

crojewsk-intel opened the pull request:
Note: this series is a sequel to #3919 and [comments](https://github.com/thesofproject/linux/pull/3919#issuecomment-1295114545) found there. Posting the entire thing, even though the avs-driver bits may be less interesting.

Patchset aims to address problem that manifests when an ASoC HDAudio driver wants to make use of reparenting functionality while still reusing HDAudio implementation found in `sound/pci/hda` which relies on stream pointer being stored in `substream->runtime->private_data`.

`dpcm_be_reparent()` manipulates and, may completely reassign BE's runtime pointer. Any data assigned in `runtime->private_data` in such case is lost. This breaks HDAudio functionality if NxFE <-> 1xBE scenario is involved.

Solution suggested here address the problem by changing location where HDAudio stream pointer is stored from `runtime->private_data` to `substream->private_data`. The latter is currently occupied by the framework - there is no location for user to store actual private data. As this information is duplicated between `pcm->private_data` and substream->private_data`, suggestion to drop usage of the latter on ASoC-framework side.

Consequently, HDAudio implementation is updated to make use of `substream->private_data` when retrieving a stream pointer. This is done by introducing `snd_substream_to_hstream()` helper macro and switching all drivers to retrieve the pointer with it.

Last two patches clean up HDAudio BackEnd operations on the avs-driver side as the new members allow for improving code cohesiveness and readability.

crojewsk-intel committed:
ASoC: Intel: avs: Refactor avs_component_hda_open()

Reverse the if-statement unlocking a possibility for lowering
indentation for the execution block.

Signed-off-by: Cezary Rojewski <cezary.rojewski@intel.com>

crojewsk-intel committed:
ASoC: Intel: avs: Use asoc_substream_to_rtd() to obtain rtd

Utilize the helper function instead of casting from ->private_data
or snd_pcm_substream_chip() directly.

Signed-off-by: Cezary Rojewski <cezary.rojewski@intel.com>

crojewsk-intel committed:
ASoC: pcm: Leave private_data of substream for the users

For ASoC users, substream->private_data is always occupied by the
framework so, there is no location to assign actual private data to.

While in several locations substream->runtime->private_data is used to
workaround this limitation, substream->runtime is manipulated by the
framework during a reparenting procedure - see dpcm_be_reparent(). To
allow users to customize their substreams and simultaneously enjoy the
reparenting functionality, use substream->pcm->private_data for
framework stuff.

Signed-off-by: Cezary Rojewski <cezary.rojewski@intel.com>

crojewsk-intel committed:
soundwire: Use asoc_substream_to_rtd() to obtain rtd

Utilize the helper function instead of casting from ->private_data
directly.

Signed-off-by: Cezary Rojewski <cezary.rojewski@intel.com>

crojewsk-intel committed:
ALSA: hda: Change location for storing HDAudio stream pointer

HDAudio PCM handling implementation expects stream pointer to be
accessible from substream->runtime->private_data whereas
substream->runtime is manipulated and may be completely reassigned by
the ASoC framework during a reparting procedure - see
dpcm_be_reparent().

To allow ASoC HDAudio drivers to utilize the reparenting functionality
while continuing to reuse ALSA HDAudio implementation, change location
where stream pointer is stored.

Signed-off-by: Cezary Rojewski <cezary.rojewski@intel.com>

crojewsk-intel committed:
ASoC: SOF: Intel: Utilize substream to HDAudio stream converters

Increase code readability and make it easier to maintain by utilizing
the new HDAudio stream accessor.

Signed-off-by: Cezary Rojewski <cezary.rojewski@intel.com>

crojewsk-intel committed:
ASoC: codecs: hda: Utilize substream to HDAudio stream converters

Increase code readability and make it easier to maintain by utilizing
the new HDAudio stream accessor.

Signed-off-by: Cezary Rojewski <cezary.rojewski@intel.com>

crojewsk-intel committed:
ALSA: hda: Introduce substream to HDAudio stream converters

To ease future refactoring, add new macros for HDAudio drivers to
retrieve stored stream pointer with.

Signed-off-by: Cezary Rojewski <cezary.rojewski@intel.com>

crojewsk-intel committed:
ASoC: Intel: avs: Utilize substream to HDAudio stream converters

Increase code readability and make it easier to maintain by utilizing
the new HDAudio stream accessor.

Signed-off-by: Cezary Rojewski <cezary.rojewski@intel.com>

crojewsk-intel committed:
ASoC: pcm: Reverse iterate DAIs when shutting them down

During startup snd_soc_dai_startup() is launched in ascending order and
the exact same thing is done during shutdown procedure. Reverse the
order in the latter so that it is symmetric to the former.

Signed-off-by: Cezary Rojewski <cezary.rojewski@intel.com>

crojewsk-intel committed:
ASoC: Intel: avs: Assign HDAudio LINK stream on DAI starutp

Streamline startup/shutdown() for HDA BE with its FE equivalent by
assigning and releasing HDAudio stream directly there.

In consequence, majority of code found in avs_hda_component_open/close()
is dropped.

Signed-off-by: Cezary Rojewski <cezary.rojewski@intel.com>

crojewsk-intel commented:
I see quite a few problems with recent upstream changes - will update. Had some conflicts when applying too. Guess it's more than it initially appeared to be. The value of the patchset won't change much though so possible review comments still apply.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -51,7 +51,7 @@ static bool hda_check_fes(struct snd_soc_pcm_runtime *rtd,
 	for_each_dpcm_fe(rtd, dir, dpcm) {
 		fe_substream = snd_soc_dpcm_get_substream(dpcm->fe, dir);
-		fe_hstream = fe_substream->runtime->private_data;
+		fe_hstream = snd_substream_to_hstream(fe_substream);
```
this isn't too good @crojewsk-intel, sometimes the helper is used to assign a "struct hdac_stream *" pointer, and ...

plbossart commented on sound/soc/intel/avs/pcm.c:
```diff
@@ -296,7 +296,7 @@ static int avs_dai_hda_be_hw_params(struct snd_pcm_substream *substream,
 		return 0;
 
-	link_stream = substream->runtime->private_data;
+	link_stream = snd_substream_to_hstream(substream);
```
... here we assign a 'struct hdac_ext_stream *' pointer.

we shouldn't add more confusion here.

crojewsk-intel replied:
I believe this is fine in C. That's how people deal with OOP concepts in the language since its beginnings.

In general, I favor the following approach:
```
struct extension {
    struct base;
    <custom field1>
    ...
    <custom fieldN>
};
```

Candidates in ALSA/ASoC for implementing such approach are many, e.g.: `struct snd_pcm_runtime` and `struct snd_soc_pcm_runtime`.

Also, I have iterated over several options. The problem being addressed here troubles any user of reparenting functionality  when HDAudio LINK stream is involved, including SOF. When fixing this, a developer has to also account for the HDAudio code found in `sound/pci/hda` and, ideally we would like to reuse it without providing any custom implementation.

plbossart replied:
No. I don't want to have to think if I am using the extended version or not. It should be very very simple and naming conventions do matter.

crojewsk-intel replied:
Would adding a second macro - `snd_substream_to_hext_stream` help? That or utilizing `stream_to_hdac_ext_stream` which is already there.

I agree on the naming part, want things clear too. First, we must take care of correctness though. A panic after `dpcm_be_reparent()` is not the right outcome.

plbossart replied:
self-explanatory macros or helpers would be fine. I just want to avoid people having to think on which layer they are using.

crojewsk-intel replied:
Introduced additional `snd_substream_to_hext_stream` macro and utilized it each time `hdac_ext_stream` instance is involved. v4 updates skylake-driver too as previous patchset revision found here ignored it.

crojewsk-intel commented:
Added missing assignment for LINK stream in `hda-dai.c`. Right now when code reaches `hdac_hda_dai_close() -> hdmi_pcm_close()` path it panics as there is no stream handle for the latter function to make use of.

Will check again tomorrow.

crojewsk-intel commented on include/sound/soc.h:
```diff
@@ -1125,6 +1125,10 @@ struct snd_soc_pcm_runtime {
+#define for_each_rtd_dais_reverse(rtd, i, dai)				\
+	for ((i) = (rtd)->dai_link->num_cpus + (rtd)->dai_link->num_codecs - 1;	\
+	     (i) >= 0 && ((dai) = (rtd)->dais[i]);			\
+	     (i)++)
```
Due to conflicts with upstream i++ got confused with i-- here. Will update next round.

crojewsk-intel commented:
We've run the recent version of the patches also with the skylake-driver. So, while area of change is delicate, we got green highway from avs, sof and skylake.

crojewsk-intel committed:
ASoC: Intel: Skylake: Utilize substream to HDAudio stream converters

Increase code readability and make it easier to maintain by utilizing
the new HDAudio stream accessor.

Signed-off-by: Cezary Rojewski <cezary.rojewski@intel.com>

crojewsk-intel commented:
With `for_each_rtd_dais_reverse()` fixed, results look very good.

Only one platform reports problems - but these seem to be connected with topology/cml_rt1011_rt5682 card enumeration.

plbossart commented on include/sound/soc.h:
```diff
@@ -1106,7 +1106,7 @@ struct snd_soc_pcm_runtime {
 #define asoc_rtd_to_codec(rtd, n) (rtd)->dais[n + (rtd)->dai_link->num_cpus]
 #define asoc_substream_to_rtd(substream) \
-	(struct snd_soc_pcm_runtime *)snd_pcm_substream_chip(substream)
+	(struct snd_soc_pcm_runtime *)(substream)->pcm->private_data
```
it's really unclear to me from the commit message what the benefits of moving the rtd storage from

substream->private_data to substream->pcm->private_data might be?

You are referring to a very obscure 'reparenting' DPCM capability but for the rest of us what does this mean?

crojewsk-intel replied:
Hmm.. I tried to be as transparent in the cover letter as possible. The case is not a common one, indeed, but has 100% reproduction rate, regardless of the HDAudio driver chosen, be it avs or sof.

The reparenting functionality is what allows ASoC to handle Nx FE <-> 1x BE and so on. Let's assume we are dealing with: 2x FE <-> 1x BE. The possibilities for open/close expand to:
```
open 1st FE
open 2nd FE
close 2nd FE
close 1st FE

open 1st FE
open 2nd FE
close 1st FE
close 2nd FE

open 2nd FE
open 1st FE
close 2nd FE
close 1st FE

open 2nd FE
open 1st FE
close 1st FE
close 2nd FE
```
Conclusion? The framework has to be able to deal with cases where FE that brought given BE to life, is closed _before_ any subsequent FE that were opened on the same BE. It effectively _reparents_ the BE to the next FE in line which is still operational. When that happens, `be_substream->runtime` pointer is **reassigned**. And with it, information previously found in `be_substream->runtime->private_data` is no longer accessible for that BE.

Now, HDAudio implementation located in `sound/pci/hda` works with `substream->runtime->private_data` containing a valid pointer to `struct hdac_stream` in mind. Combine that with reparenting and we get guaranteed panic e.g.: in `hdmi_pcm_close()` when closing one of FE connected to single BE for a HDMI endpoint.

plbossart replied:
we've removed the use of private data in SOF for the BE. see https://github.com/thesofproject/linux/pull/3908

crojewsk-intel replied:
You would have to elaborate as that does not seem true. hda-pcm.c is still full of it. More over, if you actually drop `substream->runtime->private_data` while still utilizing patches in `sound/pci/hda/patch*.c` you will get panics even if no reparenting gets involved.

plbossart replied:
sound/soc/sof/intel/hda-pcm.c is ONLY used for the FE.
sound/soc/sof/intel/hda-dai.c is ONLY used for the BE.
the BE does not rely on the private information any longer - are we in agreement on this?

plbossart commented on sound/pci/hda/hda_controller.c:
```diff
@@ -592,7 +592,7 @@ static int azx_pcm_open(struct snd_pcm_substream *substream)
 		goto unlock;
 	}
-	runtime->private_data = azx_dev;
+	substream->private_data = azx_dev;
```
You need a macro here to avoid hard-coding where the stream is stored.
You have a get, you need a set

crojewsk-intel replied:
Was thinking about it.

Agree, let's do it.

plbossart commented on sound/pci/hda/hda_controller.h:
```diff
@@ -181,7 +181,7 @@ static inline bool azx_snoop(struct azx *chip)
 static inline struct azx_dev *get_azx_dev(struct snd_pcm_substream *substream)
 {
-	return substream->runtime->private_data;
+	return substream->private_data;
```
return snd_substream_to_hstream(substream)

crojewsk-intel replied:
Ack.

plbossart commented on sound/soc/sof/intel/hda-pcm.c:
```diff
@@ -269,6 +269,6 @@ int hda_dsp_pcm_close(struct snd_sof_dev *sdev,
 
 	/* unbinding pcm substream to hda stream */
-	substream->runtime->private_data = NULL;
+	substream->private_data = NULL;
```
That's really not so good, if you really want to have a cleaner patch you should not directly handle the private_data but use helpers to store.

Actually you should first introduce the helpers to remove all directly handling of private_data, and you change the location in one shot.

crojewsk-intel replied:
The last sentence of your comment is a rocket.

Helpers first, then!

plbossart commented on sound/soc/soc-pcm.c:
```diff
@@ -712,7 +712,7 @@ static int soc_pcm_clean(struct snd_soc_pcm_runtime *rtd,
 		snd_soc_runtime_deactivate(rtd, substream->stream);
 
-	for_each_rtd_dais(rtd, i, dai)
+	for_each_rtd_dais_reverse(rtd, i, dai)
```
what does this have to do with this PR?

crojewsk-intel replied:
Well, the HDAudio stream is assigned by the platform component driver. Thus it should be released by the same driver too. Without reversing the order, codec's DAI (which are always _behind_ cpu ones) will be receiving pointer to an already released stream.

Guess that "fix" could be sent as a stand-alone patch too..
But then, the rest of the series dependent on it. It would need to be merged _before_ the rest of the pack gets sent to alsa-devel.

plbossart commented on sound/soc/intel/avs/pcm.c:
```diff
@@ -278,12 +278,33 @@ static const struct snd_soc_dai_ops avs_dai_hda_be_ops;
+
+	avs_dai_nonhda_be_shutdown(substream, dai);
+	snd_hdac_ext_stream_release(link_stream, HDAC_EXT_STREAM_TYPE_LINK);
+	substream->private_data = NULL;
```
I remember asking @ranj063 why we currently assign the link DMA in hw_params, and of course I don't recall the answer.

crojewsk-intel replied:
I believe that's because how the skylake-driver does things. LINK streams assigned in `hw_params()`, HOST streams in `open()`.

plbossart commented:
different solution pushed upstream, closing.

crojewsk-intel commented:
Wasn't expecting instant merge on upstream to be honest. Yes, after some considerations decided to stick to avs-driver-only solution for now. Was pretty slow with realizing that BE _has no private data on its own_.

I believe the `for_each_rtd_dais_reverse()` patch is still a valid upstream candidate. In regard to the rest, outcome of a another round of brainstorming is: have dynamic private data in `substream->runtime->private_data` and static one in `substream->private_data` (it is impossible at this point, yes). So, while enabling developers to utilize `substream->private_data` sounds reasonable, the HDA-reparenting issue is a reason to do so - HDAudio streams are assigned dynamically when a PCM stream opens, nothing static here.

What could be salvaged from here is addition of helpful macro - substream to hstream. @plbossart should I pursue this, or drop for now?

plbossart commented:
> What could be salvaged from here is addition of helpful macro - substream to hstream. @plbossart should I pursue this, or drop for now?

we can do this later @crojewsk-intel , we're in the middle of reshuffling the HDaudio code to handle decoupled, coupled and DMA chain options.

crojewsk-intel commented:
Resulting merge: https://lore.kernel.org/alsa-devel/20221118113052.1340593-1-cezary.rojewski@intel.com/.
